js/build/shortcode-ui.js

```diff
@@ -70,12 +71,33 @@ var MediaController = wp.media.controller.State.extend({
 	},
 
 	insert: function() {
+		var self = this;
 		var shortcode = this.props.get('currentShortcode');
-		if ( shortcode ) {
-			send_to_editor( shortcode.formatShortcode() );
-			this.reset();
-			this.frame.close();
-		}
+		var okToInsert$ = $.Deferred();
+
+		/*
+		 * Filter run before a shortcode is sent to the editor. Can be used for
+		 * client-side validation before closing the media modal.
+		 *
+		 * Called as `shortcode-ui.send_to_editor`.
+		 *
+		 * @param $.Deferred A promise which is expected to either resolve if
+		 *                   the shortcode can be sent to the editor, or reject
+		 *                   if not.
+		 * @param Shortcode  The current shortcode model.
+		 */
+		var sendToEditor$ = wp.shortcake.hooks.applyFilters( 'shortcode-ui.send_to_editor', okToInsert$, shortcode );
+
+		// Unless a filter has interfered, resolve the promise, sending the shortcode to the editor.
+		setTimeout( function() { okToInsert$.resolve(true); } );
```

Comment:
I'm not that familiar with `$.Deferred.()` and using it with `setTimeout()`, so let me try parroting this back first. If there's a hook applied, `sendToEditor$` could be a rejected promise. Otherwise it should be the value of `okToInsert$`, which is an unfulfilled promise. 

What does the timeout do? Could we set `okToInsert$` as a resolved promise up top and remove the timeout portion? Seems like it might be the same result without the timeout.